File: packages/core-ts/src/storage/machine-id.ts
import { execFile } from 'node:child_process'
import { readFile } from 'node:fs/promises'
import { hostname, networkInterfaces, userInfo } from 'node:os'
import { promisify } from 'node:util'

const execFileAsync = promisify(execFile)

export async function getMachineId(): Promise<string> {
  try {
    if (process.platform === 'linux') {
      return (await readFile('/etc/machine-id', 'utf-8')).trim()
    }
    if (process.platform === 'darwin') {
      const { stdout } = await execFileAsync('ioreg', ['-rd1', '-c', 'IOPlatformExpertDevice'])
      const match = stdout.match(/"IOPlatformUUID"\s*=\s*"([^"]+)"/)
      if (match) return match[1]
    }
    if (process.platform === 'win32') {
      // Use powershell to get MachineGuid as it's more reliable than 'reg query' parsing across encodings
      const { stdout } = await execFileAsync('powershell', [
        '-NoProfile',
        '-Command',
        'Get-ItemProperty -Path HKLM:\\SOFTWARE\\Microsoft\\Cryptography -Name MachineGuid | Select-Object -ExpandProperty MachineGuid'
      ])
      const guid = stdout.trim()
      if (guid) return guid
    }
  } catch {
    /* fallback below */
  }

  // Fallback: hostname + first MAC address
  const nics = networkInterfaces()
  const mac = Object.values(nics)
    .flat()
    .find((n) => n && !n.internal && n.mac !== '00:00:00:00:00:00')?.mac
  return `${hostname()}-${mac ?? 'unknown'}`
}

export function getUsername(): string {
  try {
    return userInfo().username
  } catch {
    return process.env.USER ?? process.env.USERNAME ?? 'unknown'
  }
}

File: packages/core-py/src/mcp_relay_core/storage/machine_id.py
"""Machine ID and username detection (cross-platform)."""

import getpass
import os
import platform
import re
import socket
import subprocess
import uuid


def get_machine_id() -> str:
    """Get a stable machine identifier.

    Platform-specific:
    - Linux: /etc/machine-id
    - macOS: IOPlatformUUID via ioreg
    - Windows: MachineGuid from registry

    Falls back to hostname + first MAC address.

    Returns:
        Machine identifier string.
    """
    system = platform.system()
    try:
        if system == "Linux":
            if os.path.exists("/etc/machine-id"):
                with open("/etc/machine-id") as f:
                    return f.read().strip()

        if system == "Darwin":
            result = subprocess.run(
                ["ioreg", "-rd1", "-c", "IOPlatformExpertDevice"],
                capture_output=True,
                text=True,
                timeout=5,
                check=False,
            )
            match = re.search(r'"IOPlatformUUID"\s*=\s*"([^"]+)"', result.stdout)
            if match:
                return match.group(1)

        if system == "Windows":
            # Use full path for reg.exe to be safe, or rely on path.
            # check=False to avoid raising CalledProcessError which we handle via Exception
            result = subprocess.run(
                [
                    "reg",
                    "query",
                    r"HKLM\SOFTWARE\Microsoft\Cryptography",
                    "/v",
                    "MachineGuid",
                ],
                capture_output=True,
                text=True,
                timeout=5,
                check=False,
            )
            match = re.search(r"MachineGuid\s+REG_SZ\s+(\S+)", result.stdout)
            if match:
                return match.group(1)
    except Exception:
        pass

    # Fallback: hostname + first MAC address
    mac = _get_first_mac()
    return f"{socket.gethostname()}-{mac}"


def _get_first_mac() -> str:
    """Get first non-loopback MAC address, or 'unknown'."""
    try:
        mac_int = uuid.getnode()
        # uuid.getnode() returns a random MAC if it can't find one,
        # indicated by the multicast bit being set
        if (mac_int >> 40) & 1:
            return "unknown"
        mac_str = ":".join(
            f"{(mac_int >> (8 * i)) & 0xFF:02x}" for i in range(5, -1, -1)
        )
        if mac_str == "00:00:00:00:00:00":
            return "unknown"
        return mac_str
    except Exception:
        return "unknown"


def get_username() -> str:
    """Get current OS username.

    Returns:
        Username string.
    """
    try:
        return getpass.getuser()
    except Exception:
        return os.environ.get("USER", os.environ.get("USERNAME", "unknown"))
